include/alpaka/mem/buf/oacc/Copy.hpp

```diff
@@ -0,0 +1,395 @@
+        struct CreateTaskMemcpy<
+            TDim,
+            DevCpu,
+            DevOacc>
+        {
+            //-----------------------------------------------------------------------------
+            template<
+                typename TExtent,
+                typename TViewSrc,
+                typename TViewDst>
+            ALPAKA_FN_HOST static auto createTaskMemcpy(
+                TViewDst & viewDst,
+                TViewSrc const & viewSrc,
+                TExtent const & extent)
+            {
+                ALPAKA_DEBUG_FULL_LOG_SCOPE;
+
+                return alpaka::oacc::detail::makeTaskCopyOacc<
+                    alpaka::oacc::detail::TaskCopyOacc,
+                    // oacc::detail::TaskCopyOacc<
+                        TDim,
+                        TViewDst,
+                        TViewSrc,
+                        TExtent>(
+                    //     void(*)(void*,void*,size_t)>(
+                            viewDst,
+                            viewSrc,
+                            extent,
+                            getDev(viewSrc),
+                            acc_memcpy_from_device
+                            );
+            }
+        };
+
+        //#############################################################################
+        //! The Oacc to Oacc memory copy trait specialization.
+        template<
+            typename TDim>
+        struct CreateTaskMemcpy<
+            TDim,
+            DevOacc,
+            DevOacc>
+        {
+            //-----------------------------------------------------------------------------
+            template<
+                typename TExtent,
+                typename TViewSrc,
+                typename TViewDst>
+            ALPAKA_FN_HOST static auto createTaskMemcpy(
+                TViewDst & viewDst,
+                TViewSrc const & viewSrc,
+                TExtent const & extent)
+            {
+                ALPAKA_DEBUG_FULL_LOG_SCOPE;
+
+                ALPAKA_ASSERT(getDev(viewDst).m_spDevOaccImpl->iDevice()
+                    == getDev(viewSrc).m_spDevOaccImpl->iDevice());
+
+                return alpaka::oacc::detail::makeTaskCopyOacc<
+                    alpaka::oacc::detail::TaskCopyOacc,
+                    // oacc::detail::TaskCopyOacc<
+                        TDim,
+                        TViewDst,
+                        TViewSrc,
+                        TExtent>(
+                    //     void(*)(void*,void*,size_t)>(
+                            viewDst,
+                            viewSrc,
+                            extent,
+                            getDev(viewDst),
+#if _OPENACC >= 201510 && ( ! defined __GNUC__ )
+                            acc_memcpy_device
+#else
+                            // acc_memcpy_device is only available since OpenACC2.5
+                            // , but we want the tests to compile anyway
+                            [](void* dst, void* src, std::size_t size){
```

Comment:
why is the memory pointed to by src not const? This makes the hacky `const_cast` above necessary.

Reply:
Because it is not const in the OpenACC API. The `const_cast` must happen somewhere.